open_flamingo/train/train_utils.py

```diff
@@ -64,193 +67,314 @@ def train_one_epoch(
+
+            model.apply(mask_embedding)
+
+            torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
+
+            # step optimizer and log
+            if (((num_steps + 1) % args.gradient_accumulation_steps) == 0) or (
+                num_steps == num_batches_per_epoch - 1
+            ):
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
 
-                wandb.log(
-                    {
-                        "data_time": data_time_m.avg,
-                        "step_time": step_time_m.avg,
-                        "laion_samples_per_second": laion_samples_per_second,
-                        "laion_samples_per_second_per_gpu": laion_samples_per_second_per_gpu,
-                        "c4_samples_per_second": c4_samples_per_second,
-                        "c4_samples_per_second_per_gpu": c4_samples_per_second_per_gpu,
-                        "lr": optimizer.param_groups[0]["lr"],
-                    },
-                    commit=False,
-                )
-                step_time_m.reset()
-                data_time_m.reset()
-
-                wandb.log(
-                    {
-                        "loss_laion": divided_loss_laion.item(),
-                        "global_step": global_step,
-                    },
-                    commit=False,
+                # step time and reset end outside of rank 0
+                step_time_m.update(time.time() - end)
+                end = time.time()
+
+                if args.rank == 0 and args.report_to_wandb:
+                    # compute within rank 0
+                    laion_samples_per_second = (
+                        args.gradient_accumulation_steps
+                        * args.batch_size_laion
+                        * args.world_size
+                        / step_time_m.val
+                    )
+                    laion_samples_per_second_per_gpu = (
+                        args.gradient_accumulation_steps
+                        * args.batch_size_laion
+                        / step_time_m.val
+                    )
+
+                    wandb.log(
+                        {
+                            "data_time": data_time_m.avg,
+                            "step_time": step_time_m.avg,
+                            "laion_samples_per_second": laion_samples_per_second,
+                            "laion_samples_per_second_per_gpu": laion_samples_per_second_per_gpu,
+                            "lr": optimizer.param_groups[0]["lr"],
+                        },
+                        commit=False,
+                    )
+                    step_time_m.reset()
+                    data_time_m.reset()
+
+                    wandb.log(
+                        {
+                            "loss_laion": loss.item(),
+                            "global_step": global_step,
+                        },
+                        commit=False,
```

Comment:
commit should be true for the last wandb log in an iteration

Reply:
Thany you for your feedback, I fixed it in next commit.